src/BoundaryConditions/fill_halo_regions.jl

```diff
@@ -22,35 +22,90 @@ function fill_halo_regions!(fields::Union{Tuple, NamedTuple}, arch, args...)
 end
 
 # Some fields have `nothing` boundary conditions, such as `FunctionField` and `ZeroField`.
-fill_halo_regions!(c::OffsetArray, ::Nothing, args...; kwargs...) = nothing
+fill_halo_regions!(c::OffsetArray, ::Nothing, args...; kwargs...) = NoneEvent()
 
 "Fill halo regions in x, y, and z for a given field's data."
-function fill_halo_regions!(c::OffsetArray, fieldbcs, arch, grid, args...; kwargs...)
+function fill_halo_regions!(c::OffsetArray, field_bcs, arch, grid, args...; kwargs...)
 
     barrier = Event(device(arch))
 
-      west_event =   fill_west_halo!(c, fieldbcs.west,   arch, barrier, grid, args...; kwargs...)
-      east_event =   fill_east_halo!(c, fieldbcs.east,   arch, barrier, grid, args...; kwargs...)
-     south_event =  fill_south_halo!(c, fieldbcs.south,  arch, barrier, grid, args...; kwargs...)
-     north_event =  fill_north_halo!(c, fieldbcs.north,  arch, barrier, grid, args...; kwargs...)
-    bottom_event = fill_bottom_halo!(c, fieldbcs.bottom, arch, barrier, grid, args...; kwargs...)
-       top_event =    fill_top_halo!(c, fieldbcs.top,    arch, barrier, grid, args...; kwargs...)
+    fill_halos! = [
+        fill_west_and_east_halo!,
+        fill_south_and_north_halo!,
+        fill_bottom_and_top_halo!,
+    ]
 
-    # Wait at the end
-    events = [west_event, east_event, south_event, north_event, bottom_event, top_event]
-    events = filter(e -> e isa Event, events)
-    wait(device(arch), MultiEvent(Tuple(events)))
+    field_bcs_array_left = [
+        field_bcs.west,
+        field_bcs.south,
+        field_bcs.bottom,
+    ]
 
-    return nothing
+    field_bcs_array_right = [
+        field_bcs.east,
+        field_bcs.north,
+        field_bcs.top,
+    ]
+
+    # perm = sortperm(field_bcs_array_left, lt=fill_first)
+    # fill_halos! = fill_halos![perm]
+    # field_bcs_array_left  = field_bcs_array_left[perm]
+    # field_bcs_array_right = field_bcs_array_right[perm]
+
+    for task = 1:3
+       fill_halo! = fill_halos![task]
+       bc_left    = field_bcs_array_left[task]
+       bc_right   = field_bcs_array_right[task]
+       events     = fill_halo!(c, bc_left, bc_right, arch, barrier, grid, args...; kwargs...)
+       
+       if eltype(events) <: Nothing
+            events = (NoneEvent(), NoneEvent())
+       end
+
+       wait(device(arch), MultiEvent(events))
+    end
+
+    return NoneEvent()
+end
+
+# Fallbacks split into two calls
+function fill_west_and_east_halo!(c, west_bc, east_bc, args...; kwargs...)
+     west_event = fill_west_halo!(c, west_bc, args...; kwargs...)
+     east_event = fill_east_halo!(c, east_bc, args...; kwargs...)
+    multi_event = (west_event, east_event)
+    return multi_event
```

Comment:
This might be an improvement:

```suggestion
    return MultiEvent((west_event, east_event))
```

and the same for those below.